src/clustering/administration/main/command_line.cc

```diff
@@ -645,6 +649,163 @@ service_address_ports_t get_service_address_ports(const std::map<std::string, op
+            return false; 
+        }
+
+        // Mutual authentication.
+        SSL_CTX_set_verify(driver_tls, SSL_VERIFY_PEER|SSL_VERIFY_FAIL_IF_NO_PEER_CERT, NULL);
+    }
+
+    return true;
+}
+
+bool configure_cluster_tls(const std::map<std::string, options::values_t> &opts, SSL_CTX *cluster_tls) {
+    boost::optional<std::string> key_file = get_optional_option(opts, "--cluster-tls-key");
+    boost::optional<std::string> cert_file = get_optional_option(opts, "--cluster-tls-cert");
+    boost::optional<std::string> ca_file = get_optional_option(opts, "--cluster-tls-ca");
+
+    if (!(key_file && cert_file && ca_file)) {
+        logERR("must specify each of --cluster-tls-key, --cluster-tls-cert, and --cluster-tls-ca");
+        return false;
+    }
+
+    if (!load_tls_key_and_cert(cluster_tls, *key_file, *cert_file)) {
+        return false;
+    }
+
+    if (!SSL_CTX_load_verify_locations(cluster_tls, ca_file->c_str(), NULL)) {
+        ERR_print_errors_fp(stderr);
+        return false; 
+    }
+
+    // Mutual authentication.
+    SSL_CTX_set_verify(cluster_tls, SSL_VERIFY_PEER|SSL_VERIFY_FAIL_IF_NO_PEER_CERT, NULL);
+
+    return true;
+}
+
+bool initialize_tls_ctx(SSL_CTX **tls_ctx) {
+    *tls_ctx = SSL_CTX_new(SSLv23_method());
+    if (NULL == *tls_ctx) {
+        ERR_print_errors_fp(stderr);
+        return false;
+    }
+
+    // Only allow TLS v1.2 and prefer server ciphers.
+    SSL_CTX_set_options(*tls_ctx, SSL_OP_NO_SSLv2|SSL_OP_NO_SSLv3|SSL_OP_NO_TLSv1|SSL_OP_NO_TLSv1_1|SSL_OP_CIPHER_SERVER_PREFERENCE);
+
+    /* This is pretty important. We want to use the most secure TLS cipher
+    suite that we can. This simple list only allows ciphers suites which employ
+    ECDHE (Elliptic Curve Diffie-Hellman with Ephemeral keys) for encryption
+    key agreement for Perfect Forward Secrecy and we want to use AESGCM for
+    efficient authenticated encryption. */
+    if (0 == SSL_CTX_set_cipher_list(*tls_ctx, "ECDHE+AESGCM")) {
+        logNTC("No secure cipher suites available\n");
+        return false;
+    }
+
+    /* These values correspond to the standard NIST/FIPS curves and are also
+    commonly known as P-521, P-384, and P-256 respectively. Some people
+    speculate that these curves are compromised in some way by the NSA but most
+    people agree that they are safe to use. Any real risk in using elliptic
+    curves is associated with ECDSA (Elliptic Curve Digital Signature
+    Algorithm) where you must be careful to generate signatures correctly using
+    a secure PRNG. As we are using them only for ECDHE, I stand by this
+    selection. Feel free to add any more that we want - for example 'secp256k1'
+    (widely used and popularized by the Bitcoin cryptocurrency) and Daniel J.
+    Bernstein's 'Curve25519' (though not currently supported by OpenSSL) would
+    be other good choices. */
+    if (0 == SSL_CTX_set1_curves_list(*tls_ctx, "secp521r1:secp384r1:prime256v1")) {
```

Comment:
I'd love to pick Curve25519 and call it a day, but unfortunately that isn't an option. That said, this set looks like the best we can do for now.

Similarly it might be nice to allow this to be configurable for when Curve25519 becomes available. It might already be if you're using a drop-in replacement like BoringSSL or LibreSSL.